File: 160.Intersection-of-Two-Linked-Lists.py
# https://leetcode.com/problems/intersection-of-two-linked-lists/
#
# algorithms
# Easy (31.19%)
# Total Accepted:    254,398
# Total Submissions: 815,597


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        tmp_a, tmp_b = headA, headB
        len_a, len_b = 0, 0

        while tmp_a:
            len_a += 1
            tmp_a = tmp_a.next

        while tmp_b:
            len_b += 1
            tmp_b = tmp_b.next

        if len_a == 0 or len_b == 0:
            return None

        if len_a > len_b:
            while len_a > len_b:
                headA = headA.next
                len_a -= 1
        elif len_b > len_a:
            while len_b > len_a:
                headB = headB.next
                len_b -= 1

        while headA and headB:
            if headA == headB:
                return headA
            headA = headA.next
            headB = headB.next

        return None

File: 160.Intersection-of-Two-Linked-Lists.java
// https://leetcode.com/problems/intersection-of-two-linked-lists/
//
// algorithms
// Easy (33.21%)
// Total Accepted:    291,291
// Total Submissions: 877,232
// beats 100.0% of java submissions


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int ALen = 0, BLen = 0;
        ListNode tmp = headA;
        
        while (tmp != null) {
            ALen++;
            tmp = tmp.next;
        }
        
        tmp = headB;
        
        while (tmp != null) {
            BLen++;
            tmp = tmp.next;
        }
        
        ListNode fast = (ALen >= BLen) ? headA : headB;
        ListNode slow = (ALen >= BLen) ? headB : headA;
        for (int i = 0; i < Math.abs(ALen - BLen); i++) {
            fast = fast.next;
        }
        
        while (fast != null && slow != null) {
            if (fast == slow) {
                return fast;
            }
            fast = fast.next;
            slow = slow.next;
        }
        
        return null;
    }
}